### PR TITLE
Don't use match-destructuring for derived ops on structs.

### PR DESCRIPTION
**Pull request by nnethercote**
r? @scottmcm

### CODE REVIEW

**Comment by nnethercote** [comment]
Please note!
- Best reviewed one commit at a time.
- The first six commits are from #98376, which already have r+ and don't need reviewing. The last four commits build on them.

**Comment by nnethercote** [comment]
We'll need to wait for #98376 to land and the a rebase before doing a perf run here, otherwise the perf improvements from that PR will be combined with the perf improvements from this PR. But I did local measurements and saw instruction count reductions of up to 17% on the `derive` stress test and up to 3% on primary benchmarks.

**Comment by scottmcm** [comment]
(Haven't looked in detail yet, but reading the commit message made me think)

If you need something short that works for packed structs, could you maybe use [`&{self.x}`](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=c10e756e8e32848dfbfcde04d108902a)?  I think [that](https://bluss.github.io/rust/fun/2015/10/11/stuff-the-identity-function-does/#rust-has-dedicated-syntax-for-this) still gets the move-then-ref-the-temporary that's needed for soundness -- it certainly doesn't trigger the warning -- but using way fewer tokens and no inventing of new names.

**Comment by nnethercote** [comment]
> could you maybe use [&{self.x}](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=c10e756e8e32848dfbfcde04d108902a)

Oh neat! That will let me get rid of all the pattern and `let` stuff. Thanks for the tip, I'll do it on Monday.

File: src/test/ui/deriving/deriving-all-codegen.rs
Comment (AngelicosPhosphoros) [inline src/test/ui/deriving/deriving-all-codegen.rs:67]: It is a enum and you missing non-Copy struct, actually.

Comment (nnethercote) [reply src/test/ui/deriving/deriving-all-codegen.rs:67]: Good catch. I will fix it on Monday.

**Comment by nnethercote** [comment]
> We'll need to wait for #98376 to land and the a rebase before doing a perf run here, otherwise the perf improvements from that PR will be combined with the perf improvements from this PR.

Actually, there's no harm in seeing a combined measurement:

@bors try @rust-timer queue

**Comment by scottmcm** [comment]
It's so strange to me that `regex opt full` gets worse, but `regex opt incr-full` gets (negligibly) better.

**Comment by nnethercote** [comment]
> It's so strange to me that `regex opt full` gets worse, but `regex opt incr-full` gets (negligibly) better.

I have a possible explanation [here](https://github.com/rust-lang/rust/pull/98190#issuecomment-1162617774). That PR hasn't landed yet, hopefully it's effectively the same regression we're seeing, and once #98180 is merged (another precursor to this PR) the regression will disappear here...

File: src/test/ui/deriving/deriving-all-codegen.rs
Comment (nnethercote) [reply src/test/ui/deriving/deriving-all-codegen.rs:67]: I ended up removing this struct in #98376, which this PR depends on.

**Comment by nnethercote** [comment]
I tried the `&{self.x}` approach and it's 99% great but falls down on 1% of cases. Consider this code:
```rust
#[derive(Hash)]
struct A;

#[derive(Hash)]
#[repr(packed)]
struct Foo(A);

#[derive(Hash)]
#[repr(packed)]
pub struct Bar<T: Copy>(T);

fn main() {}
```
Currently, rustc accepts `A` and `Foo` but rejects `Bar`:
```
error: `#[derive]` can't be used on a `#[repr(packed)]` struct with type or const parameters (error E0133)
 --> f.rs:8:10
  |
8 | #[derive(Hash)]
  |          ^^^^
  |
  = note: `#[deny(unaligned_references)]` on by default
  = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
  = note: for more information, see issue #82523 <https://github.com/rust-lang/rust/issues/82523>
  = note: this error originates in the derive macro `Hash` (in Nightly builds, run with -Z macro-backtrace for more info)
```
The error on `Bar` is triggered by a check done on MIR.

Using `&{self.x}`, rustc accepts `A` and `Bar` but rejects `Foo`:
```
error[E0507]: cannot move out of `self.0` which is behind a shared reference
 --> f.rs:6:12
  |
4 | #[derive(Hash)]
  |          ---- in this derive macro expansion
5 | #[repr(packed)]
6 | struct Foo(A);
  |            ^ move occurs because `self.0` has type `A`, which does not implement the `Copy` trait
  |
  = note: this error originates in the derive macro `Hash` (in Nightly builds, run with -Z macro-backtrace for more info)
```
The error on `Foo` is because the `&{self.x}` relies on `self.x` implementing `Copy`, which isn't true here. It's also interesting that `Bar` is accepted -- it shows that error E0133 is possibly stricter than it needs to be.

These are just two examples, there are some other minor variations. `src/test/ui/derives/deriving-with-repr-packed.rs` is a test demonstrating some of these.

Given all that, I will stick to the well-tested let-pattern approach. Thanks again for the suggestion!

**Comment by nnethercote** [comment]
The final four commits here are ready for review.

**Comment by nnethercote** [comment]
@scottmcm: I have rebased, and now all three commits are ready for review. Thanks!

File: compiler/rustc_builtin_macros/src/deriving/generic/mod.rs
Comment (scottmcm) [inline compiler/rustc_builtin_macros/src/deriving/generic/mod.rs:1029]: I'm unsure exactly when `use_temporaries` is needed now that `is_packed` seems to be making many of the relevant decisions.  I tried to follow it down a bit more, but got a bit lost.

Are there any cases that strictly *need* temporaries when `is_packed` is false?

(I don't think this needs to block this PR, but might be worth exploring in future cleanup, especially if it could make callsites clearer to have an enum for the cases, rather than the bunch-of-`bool`s that some of them seem to have today.)

File: compiler/rustc_expand/src/build.rs
Comment (scottmcm) [inline compiler/rustc_expand/src/build.rs:158]: fix: seems like this parameter should just be removed, then?  It'll be part of the `Pat` in any case that needs it.

(Conveniently all the other parameters have non-primitive types, so consistency with `stmt_let` doesn't seem worth it to me.)

File: compiler/rustc_builtin_macros/src/deriving/generic/mod.rs
Comment (nnethercote) [reply compiler/rustc_builtin_macros/src/deriving/generic/mod.rs:1029]: `use_temporaries` is still used in `create_struct_pattern`, which has two call sites, so the `is_packed`/`use_temporaries` distinction is still required. I haven't changed this in any of my follow-ups, but I'll think about if it can be improved.

File: src/test/ui/deriving/deriving-all-codegen.stdout
Comment (scottmcm) [inline src/test/ui/deriving/deriving-all-codegen.stdout:421]: curiosity, no changes needed: very odd that these end up formatted with decreasing indents, like this.

File: compiler/rustc_expand/src/build.rs
Comment (nnethercote) [reply compiler/rustc_expand/src/build.rs:158]: Good catch. I think I copied `stmt_let_ty` to get the starting point for this function and didn't remove this unused argument. I will fix it shortly.

File: src/test/ui/deriving/deriving-all-codegen.stdout
Comment (nnethercote) [reply src/test/ui/deriving/deriving-all-codegen.stdout:421]: Yeah, I assume it's the pretty-printer doing a not great job.

**Comment by scottmcm** [comment]
Looks good!  r=me with the one unused parameter removed (https://github.com/rust-lang/rust/pull/98446#discussion_r911495731).

**Comment by scottmcm** [comment]
No idea why your comment didn't take.  It's not showing in the queue either.

@bors r+

**Comment by nnethercote** [comment]
@richkadel This PR changes the code generated for derived code for various built-in traits. It's causing some changes in coverage test output.

I ran `x test src/test/run-make-fulldeps/coverage-reports --bless` to generate the 4th commit. The changes just seem like minor perturbations -- are they important? I don't understand these tests at all so I can't tell if the changes are reasonable.

**Comment by nnethercote** [comment]
Also, this is the second time I've hit a change in the coverage test output, and I have numerous more changes to the derive builtins in progress and planned. Is there any documentation I can read on these tests to help me understand any other future changes I might hit? Thanks.

**Comment by richkadel** [comment]
> Also, this is the second time I've hit a change in the coverage test output, and I have numerous more changes to the derive builtins in progress and planned. Is there any documentation I can read on these tests to help me understand any other future changes I might hit? Thanks.

It's very unusual for the coverage to change, but when it does I just try to verify the logic in it. For instance if a coverage changes from 0 to 1 on derived trait that should mean that we're counting the execution of a trait in that derived trait that we weren't counting before, which should mean that it is executing now, and was not executing before. I'm on my mobile device because I'm on vacation, so I may have a hard time seeing the changes completely, but it looks like some counters are lost in your new version, which seems like a problem to me. There can be differences in the ordering of results in reports, especially if you are regenerating the results on Windows. Unix variants, including Linux and Mac OS, are generally very consistent in my experience. If you are dropping coverage counters, that may indicate a problem. I wouldn't expect that to happen. There is some fairly detailed documentation on the rust coverage implementation in the developer guide. 

I can take a closer look at this when I return from vacation in a few days.

File: src/test/run-make-fulldeps/coverage-reports/expected_show_coverage.uses_crate.txt
Comment (richkadel) [inline src/test/run-make-fulldeps/coverage-reports/expected_show_coverage.uses_crate.txt:39]: I think these are just reordered. If you're generating the new coverage results on Windows, that's a common problem but it's not wrong. Sadly llvm coverage report generation does not have deterministic ordering of results. Generally they are very consistently ordered on Linux specifically, and just as consistently ordered in the opposite way on Windows. I'm not sure if that's the issue here or not.

**Review by richkadel — CHANGES_REQUESTED** [review]

File: src/test/run-make-fulldeps/coverage-reports/expected_show_coverage.partial_eq.txt
Comment (nnethercote) [reply src/test/run-make-fulldeps/coverage-reports/expected_show_coverage.partial_eq.txt:5]: Why does the original output have seven `^0`/`^1` fields here when there are only six derived traits?

Comment (nnethercote) [reply src/test/run-make-fulldeps/coverage-reports/expected_show_coverage.partial_eq.txt:5]: And are the `^0`/`^1` fields supposed to line up with the start of the trait name? Some do, some don't.

**Comment by nnethercote** [comment]
Good news: fixing up a single span in the generated code was enough to revert all the coverage test changes except for the `&str`/`alloc::vec::Vec<i32>` ordering one, and it makes sense that this is a trivial change due to the nondeterminism.

**Comment by richkadel** [comment]
> Good news: fixing up a single span in the generated code was enough to revert all the coverage test changes except for the `&str`/`alloc::vec::Vec<i32>` ordering one, and it makes sense that this is a trivial change due to the nondeterminism.

That's great news!

**Pull request merged**
